File: packages/api/go/core/services/mediator/sql_external_mediator.go
package mediator

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"

	"template/core-go/types"
)

// ════════════════════════ THE OUTBOX IS THE TRANSPORT ════════════════════════
//
// SqlExternalMediator is the concrete ExternalMediator that replaces the Redis
// Streams mediator (go-domain-design.md §1.3, decisions §3(b)/§3(c)). Instead of
// outbox → Redis XADD → XREADGROUP, the consumer reads the outbox rows DIRECTLY:
// the shared SQLite database (WAL) carries integration events between processes
// and, in the single-binary target, between contexts of one process.
//
// Delivery is at-least-once: a crash after dispatch but before MarkProcessed
// redelivers. Exactly-once is NOT a transport property — it comes from the
// destination's idempotency (idempotent handler + a dedup UNIQUE such as
// thread.consumed_messages + INSERT ... ON CONFLICT DO NOTHING). at-least-once +
// dedup UNIQUE = exactly-once.
//
// Two roles, cleanly split (mirrors the retired Redis egress semantics):
//   - Publish (egress): INSERT the event as an outbox row (durable transport) and
//     synchronously fan out to in-process callbacks — the SSE broadcaster in
//     listen_events.go sees the event immediately, with no poll latency. Then
//     best-effort Notify the strategy to wake the consumer.
//   - the consume loop (ingress): claim unprocessed outbox rows via a leased
//     UPDATE and dispatch them to registered handlers. With no ingress handlers
//     registered (the channel gateway is egress-only today) the loop is a no-op
//     consumer — it claims nothing, so egress rows are never dropped.
const (
	// integrationClaimBatch bounds one claim cycle.
	integrationClaimBatch = 50
	// integrationLease is how long a claimed row is hidden from other consumers
	// (and from this consumer's own retry) before its lease expires and it becomes
	// re-claimable. It doubles as the retry backoff for a failed dispatch.
	integrationLease = 30 * time.Second
	// integrationMaxAttempts dead-letters a row after this many failed dispatches.
	integrationMaxAttempts = 5
)

// outboxRow is a single claimed integration-event outbox row.
type outboxRow struct {
	id       string
	name     string
	ownerID  string
	payload  json.RawMessage
	attempts int64
}

// SqlExternalMediator implements ExternalMediator over the SQLite outbox.
type SqlExternalMediator struct {
	db     *sql.DB
	source string
	notify NotifyStrategy

	mu        sync.RWMutex
	handlers  map[string][]IntegrationEventHandler
	callbacks []func(ctx context.Context, event types.IntegrationEventI)

	started bool
	cancel  context.CancelFunc
	done    chan struct{}

	// now is the injectable clock (tests drive lease expiry deterministically).
	now func() time.Time
}

// NewSqlExternalMediator builds the mediator. source is this producer's outbox
// discriminator (the "source" column); the consume loop claims only rows of this
// source whose name has a registered handler. notify decides how the consumer
// wakes (in-process nudge and/or WAL poll — see SqliteWalPollingStrategy).
func NewSqlExternalMediator(db *sql.DB, source string, notify NotifyStrategy) *SqlExternalMediator {
	return &SqlExternalMediator{
		db:       db,
		source:   source,
		notify:   notify,
		handlers: make(map[string][]IntegrationEventHandler),
		done:     make(chan struct{}),
		now:      time.Now,
	}
}

// Register records an ingress handler. Unlike the retired RedisExternalMediator
// (which failed loud because it had no consumer), the SQL mediator's whole point
// is a working consumer — so this is a live registration the claim loop honors.
func (m *SqlExternalMediator) Register(h IntegrationEventHandler) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.handlers[h.EventName()] = append(m.handlers[h.EventName()], h)
	slog.Info("sql external mediator: registered handler", "event", h.EventName())
}

// RegisterCallback records an in-process fan-out callback (e.g. the SSE broadcaster).
func (m *SqlExternalMediator) RegisterCallback(fn func(ctx context.Context, event types.IntegrationEventI)) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.callbacks = append(m.callbacks, fn)
}

// Publish inserts the event as an outbox row (the durable transport), fans it out
// to local callbacks, and nudges the consumer. The INSERT is idempotent
// (ON CONFLICT(id) DO NOTHING) so a re-published event id is a no-op rather than a
// PK error.
func (m *SqlExternalMediator) Publish(ctx context.Context, event types.IntegrationEventI) error {
	payload, err := json.Marshal(event)
	if err != nil {
		return fmt.Errorf("sql external mediator: marshal event %s: %w", event.GetEventName(), err)
	}

	// ids are TEXT in SQLite — plain `?` bindings, no `::uuid` cast (the pg outbox
	// uuid-cast discipline does not carry to this dialect).
	if _, err := m.db.ExecContext(ctx,
		`INSERT INTO shared_outbox (id, name, entity_id, owner_id, payload, source, created_at)
		 VALUES (?, ?, NULL, ?, ?, ?, ?)
		 ON CONFLICT(id) DO NOTHING`,
		integrationEventID(event),
		event.GetEventName(),
		nullableText(event.GetOwnerID()),
		string(payload),
		m.source,
		m.now().UnixMilli(),
	); err != nil {
		return fmt.Errorf("sql external mediator: insert outbox row for %s: %w", event.GetEventName(), err)
	}

	m.fireCallbacks(ctx, event)

	if m.notify != nil {
		_ = m.notify.Notify(ctx, m.source)
	}
	return nil
}

// Start launches the consume loop. Idempotent: a second Start while running is a
// no-op (the memoized composition root may resolve the mediator more than once).
func (m *SqlExternalMediator) Start(_ context.Context) error {
	m.mu.Lock()
	if m.started {
		m.mu.Unlock()
		return nil
	}
	m.started = true
	// The loop owns its own context so it outlives the fx OnStart ctx; Stop cancels it.
	ctx, cancel := context.WithCancel(context.Background())
	m.cancel = cancel
	m.mu.Unlock()

	go m.consumeLoop(ctx)
	slog.Info("sql external mediator: started", "source", m.source)
	return nil
}

// Stop cancels the consume loop, waits for it to drain out, and closes the notify
// strategy. Safe to call once.
func (m *SqlExternalMediator) Stop(_ context.Context) error {
	m.mu.Lock()
	if !m.started {
		m.mu.Unlock()
		return nil
	}
	m.started = false
	cancel := m.cancel
	m.mu.Unlock()

	if cancel != nil {
		cancel()
	}
	<-m.done
	if m.notify != nil {
		_ = m.notify.Close()
	}
	slog.Info("sql external mediator: stopped", "source", m.source)
	return nil
}

// consumeLoop drains all currently-claimable work on each wake, then blocks on the
// NotifyStrategy until the next nudge or poll tick.
func (m *SqlExternalMediator) consumeLoop(ctx context.Context) {
	defer close(m.done)
	for {
		for {
			n, err := m.drainOnce(ctx)
			if err != nil {
				if ctx.Err() != nil {
					return
				}
				slog.Error("sql external mediator: drain error", "source", m.source, "error", err)
				break
			}
			if n < integrationClaimBatch {
				break // fewer than a full batch → nothing left this cycle
			}
		}
		if err := m.notify.Wait(ctx, m.source); err != nil {
			return // ctx cancelled
		}
	}
}

// drainOnce claims a batch and dispatches it. Returns the number of rows claimed.
func (m *SqlExternalMediator) drainOnce(ctx context.Context) (int, error) {
	names := m.handlerNames()
	if len(names) == 0 {
		return 0, nil // egress-only: no ingress handlers → claim nothing, drop nothing
	}

	rows, err := m.claim(ctx, names)
	if err != nil {
		return 0, err
	}
	for _, r := range rows {
		m.process(ctx, r)
	}
	return len(rows), nil
}

// claim leases up to integrationClaimBatch unprocessed rows (of this source, whose
// name a handler is registered for) under a single write transaction — the §3(c)
// claim: stamp claimed_by (an opaque per-claim token) + lease_until (now + lease),
// then read back exactly the rows this token won. NO consumer-groups: a single
// consumer per direction, redelivery deduped at the destination.
//
// The transaction's first statement is the UPDATE (a write), so SQLite takes the
// write lock at that point — equivalent to BEGIN IMMEDIATE for our purpose (no
// deferred-then-upgrade deadlock window).
func (m *SqlExternalMediator) claim(ctx context.Context, names []string) ([]outboxRow, error) {
	token := uuid.NewString()
	now := m.now().UnixMilli()
	leaseUntil := now + integrationLease.Milliseconds()

	tx, err := m.db.BeginTx(ctx, nil)
	if err != nil {
		return nil, fmt.Errorf("begin claim tx: %w", err)
	}
	committed := false
	defer func() {
		if !committed {
			_ = tx.Rollback()
		}
	}()

	placeholders := make([]string, len(names))
	for i := range names {
		placeholders[i] = "?"
	}
	nameIn := strings.Join(placeholders, ", ")

	// UPDATE ... WHERE id IN (SELECT ... LIMIT n): the leased claim.
	updateArgs := make([]any, 0, len(names)+4)
	updateArgs = append(updateArgs, token, leaseUntil, m.source, now)
	for _, n := range names {
		updateArgs = append(updateArgs, n)
	}
	updateArgs = append(updateArgs, integrationClaimBatch)

	if _, err = tx.ExecContext(ctx, fmt.Sprintf(
		`UPDATE shared_outbox
		 SET claimed_by = ?, lease_until = ?
		 WHERE id IN (
		     SELECT id FROM shared_outbox
		     WHERE source = ?
		       AND processed_at IS NULL
		       AND (lease_until IS NULL OR lease_until < ?)
		       AND name IN (%s)
		     ORDER BY created_at
		     LIMIT ?
		 )`, nameIn), updateArgs...); err != nil {
		return nil, fmt.Errorf("claim update: %w", err)
	}

	rows, err := tx.QueryContext(ctx,
		`SELECT id, name, owner_id, payload, attempts
		 FROM shared_outbox
		 WHERE claimed_by = ?
		 ORDER BY created_at`, token)
	if err != nil {
		return nil, fmt.Errorf("select claimed rows: %w", err)
	}

	var claimed []outboxRow
	for rows.Next() {
		var (
			r       outboxRow
			owner   sql.NullString
			payload string
		)
		if scanErr := rows.Scan(&r.id, &r.name, &owner, &payload, &r.attempts); scanErr != nil {
			rows.Close()
			return nil, fmt.Errorf("scan claimed row: %w", scanErr)
		}
		r.ownerID = owner.String
		r.payload = json.RawMessage(payload)
		claimed = append(claimed, r)
	}
	rows.Close()
	if err = rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate claimed rows: %w", err)
	}

	if err = tx.Commit(); err != nil {
		return nil, fmt.Errorf("commit claim tx: %w", err)
	}
	committed = true
	return claimed, nil
}

// process dispatches one claimed row to its handlers, then marks it processed. On
// dispatch failure it records the error and either dead-letters (>= max attempts)
// or leaves the lease in place so the row is retried after the lease expires
// (natural backoff — no hot-loop on a poison message).
func (m *SqlExternalMediator) process(ctx context.Context, r outboxRow) {
	event := rawIntegrationEventFromRow(r)
	if err := m.dispatch(ctx, event); err != nil {
		m.fail(ctx, r, err)
		return
	}
	if err := m.markProcessed(ctx, r.id); err != nil {
		// The row keeps its lease; it will be redelivered after expiry. Handlers are
		// idempotent (dedup UNIQUE at destination), so a duplicate delivery is safe.
		slog.Error("sql external mediator: mark processed failed",
			"id", r.id, "event", r.name, "error", err)
	}
}

// dispatch runs every handler registered for the event, with panic recovery,
// returning the first error so the row is retried.
func (m *SqlExternalMediator) dispatch(ctx context.Context, event types.IntegrationEventI) error {
	m.mu.RLock()
	handlers := append([]IntegrationEventHandler(nil), m.handlers[event.GetEventName()]...)
	m.mu.RUnlock()

	for _, h := range handlers {
		if err := m.safeHandle(ctx, h, event); err != nil {
			return err
		}
	}
	return nil
}

func (m *SqlExternalMediator) safeHandle(ctx context.Context, h IntegrationEventHandler, event types.IntegrationEventI) (err error) {
	defer func() {
		if r := recover(); r != nil {
			slog.Error("sql external mediator: handler panicked",
				"event", event.GetEventName(), "handler", fmt.Sprintf("%T", h), "panic", r)
			err = fmt.Errorf("handler panicked: %v", r)
		}
	}()
	return h.Handle(ctx, event)
}

func (m *SqlExternalMediator) markProcessed(ctx context.Context, id string) error {
	_, err := m.db.ExecContext(ctx,
		`UPDATE shared_outbox SET processed_at = ?, claimed_by = NULL WHERE id = ?`,
		m.now().UnixMilli(), id)
	return err
}

func (m *SqlExternalMediator) fail(ctx context.Context, r outboxRow, cause error) {
	attempts := r.attempts + 1
	if attempts >= integrationMaxAttempts {
		slog.Warn("sql external mediator: dead-lettering event",
			"id", r.id, "event", r.name, "attempts", attempts, "error", cause)
		// processed_at set → stops being claimed; last_error kept for audit.
		if _, err := m.db.ExecContext(ctx,
			`UPDATE shared_outbox SET attempts = ?, last_error = ?, processed_at = ?, claimed_by = NULL WHERE id = ?`,
			attempts, cause.Error(), m.now().UnixMilli(), r.id); err != nil {
			slog.Error("sql external mediator: dead-letter update failed", "id", r.id, "error", err)
		}
		return
	}
	slog.Error("sql external mediator: dispatch failed, will retry after lease",
		"id", r.id, "event", r.name, "attempts", attempts, "error", cause)
	// Keep claimed_by/lease_until in place → the row is hidden until the lease
	// expires, then reclaimed. attempts + last_error advance for observability.
	if _, err := m.db.ExecContext(ctx,
		`UPDATE shared_outbox SET attempts = ?, last_error = ? WHERE id = ?`,
		attempts, cause.Error(), r.id); err != nil {
		slog.Error("sql external mediator: retry update failed", "id", r.id, "error", err)
	}
}

// handlerNames returns the distinct event names with at least one registered
// handler — the claim's name filter (an empty set means egress-only).
func (m *SqlExternalMediator) handlerNames() []string {
	m.mu.RLock()
	defer m.mu.RUnlock()
	names := make([]string, 0, len(m.handlers))
	for name, hs := range m.handlers {
		if len(hs) > 0 {
			names = append(names, name)
		}
	}
	return names
}

func (m *SqlExternalMediator) fireCallbacks(ctx context.Context, event types.IntegrationEventI) {
	m.mu.RLock()
	callbacks := make([]func(ctx context.Context, event types.IntegrationEventI), len(m.callbacks))
	copy(callbacks, m.callbacks)
	m.mu.RUnlock()
	for _, fn := range callbacks {
		fn(ctx, event)
	}
}

// integrationEventID prefers the event's own deterministic id (types.IntegrationEvent[T]
// exposes GetEventID) so a re-publish maps onto the same outbox row; otherwise it
// mints a fresh one.
func integrationEventID(event types.IntegrationEventI) string {
	if p, ok := event.(interface{ GetEventID() string }); ok {
		if id := p.GetEventID(); id != "" {
			return id
		}
	}
	return uuid.NewString()
}

func nullableText(s string) any {
	if s == "" {
		return nil
	}
	return s
}

// integrationEventEnvelope mirrors the JSON written by json.Marshal(types.IntegrationEvent[T]).
type integrationEventEnvelope struct {
	ID      string          `json:"id"`
	OwnerID string          `json:"ownerId"`
	Name    string          `json:"name"`
	Payload json.RawMessage `json:"payload"`
}

// rawIntegrationEvent is a non-generic IntegrationEventI rebuilt from an outbox
// row, exposing the inner payload via GetPayload so UnmarshalIntegrationEvent[T]
// can extract a typed payload — the same slow-path shape rawDomainEvent uses.
type rawIntegrationEvent struct {
	id      string
	ownerID string
	name    string
	payload json.RawMessage
}

func (e *rawIntegrationEvent) GetEventName() string        { return e.name }
func (e *rawIntegrationEvent) GetOwnerID() string          { return e.ownerID }
func (e *rawIntegrationEvent) GetEventID() string          { return e.id }
func (e *rawIntegrationEvent) GetPayload() json.RawMessage { return e.payload }

// rawIntegrationEventFromRow reconstructs the event from the stored envelope,
// falling back to the row-level columns when the envelope is absent/partial.
func rawIntegrationEventFromRow(r outboxRow) *rawIntegrationEvent {
	var env integrationEventEnvelope
	if err := json.Unmarshal(r.payload, &env); err != nil {
		return &rawIntegrationEvent{id: r.id, ownerID: r.ownerID, name: r.name, payload: r.payload}
	}
	id := env.ID
	if id == "" {
		id = r.id
	}
	ownerID := env.OwnerID
	if ownerID == "" {
		ownerID = r.ownerID
	}
	name := env.Name
	if name == "" {
		name = r.name
	}
	payload := env.Payload
	if len(payload) == 0 {
		payload = r.payload
	}
	return &rawIntegrationEvent{id: id, ownerID: ownerID, name: name, payload: payload}
}

// Compile-time assertions.
var (
	_ ExternalMediator        = (*SqlExternalMediator)(nil)
	_ types.IntegrationEventI = (*rawIntegrationEvent)(nil)
)

File: packages/api/typescript/core/src/services/Mediator/SqlExternalMediator.ts
import { injectable } from 'tsyringe-neo'
import { eq, inArray, sql } from 'drizzle-orm'
import { outbox } from '@codm/contracts/db'
import { BaseInfrastructureErrors } from '../../errors/codes'
import { BaseEvent } from '../../types/BaseEvent'
import { BaseError } from '../../types/BaseError'
import { Handler } from '../../types/Handler'
import { tryCatchAsync } from '../../utils/TryCatch'
import { LibSqlDatabaseDriver } from '../../db/libsql/LibSqlDatabaseDriver'
import { DomainEventRepository } from '../../repositories/DomainEventRepository'
import type { PollingService } from '../HealthService/HealthCheck'
import { BaseIntegrationEvent, type AnyIntegrationEvent } from '../../types/BaseIntegrationEvent'
import { EventCallback, ExternalMediator, Unsubscribe, handlerEventNames } from './Mediator'
import { adaptWireEnvelope, reviveIsoDates } from './wire'

/**
 * The lane this mediator owns — and it carries BOTH directions.
 *
 * The old text described `integration` as the Go gateway's egress to us and nothing else. That was a
 * restriction of the moment, not a property of the lane (founder, 29-jul): since B3 the TS side
 * PUBLISHES here too (`publish()` → `saveIntegrationEvent`), and the claim does not care who produced
 * a row — it filters by NAME (only names with a registered handler) and leases. There is still exactly
 * ONE claimant, because the Go twin is built EGRESS-ONLY on this lane
 * (`NewSqlExternalMediatorWithoutIngress`); when Go one day needs to consume a TS-published fact it
 * registers a handler for that name — same lane, same claim, no new lane.
 * (`api` belongs to LibSqlOutboxDispatcher, `gateway` to the Go dispatcher.)
 */
const LANE = 'integration'

const BATCH_SIZE = 50
const MAX_ATTEMPTS = 5
const LEASE_MS = 30_000

// Poll bounds. The MAX is 2s, not the domain dispatcher's 30s, mirroring the Go WAL polling
// strategy: SQLite has no cross-process push, and a `integration.channel.connected` sitting half a
// minute in the table reads to the user as exactly the DISCONNECTED console this phase exists to
// kill.
const POLL_MIN_MS = 50
const POLL_MAX_MS = 2_000
const POLL_BACKOFF_FACTOR = 1.5

interface ClaimedOutboxRow {
	id: string
	name: string
	owner_id: string | null
	/** RAW TEXT — parsed here, with the date reviver. Never through the ORM's json column mode. */
	payload: string | null
	attempts: number
}

/**
 * Shared-outbox ingress: the TS twin of the Go `SqlExternalMediator`.
 *
 * TRANSPORT. There is no broker and no socket — the transport IS the shared SQLite file. The
 * gateway INSERTs a row with `source = 'integration'`; this class claims it by lease, revives it,
 * and hands it to the registered external handlers. That is also why it can be the `real` binding
 * even under the e2e harness, which boots no Go process: nothing about it needs a peer to be up.
 *
 * PUBLISH PERSISTS, AND ONLY PERSISTS (B3, decisions 4/5). `publish()` INSERTs the event on this lane
 * via `DomainEventRepository.saveIntegrationEvent` and returns — it dispatches nothing in the caller's
 * call stack and fires no callback there. Every delivery, Go-published or TS-published, comes from
 * `drainOnce`: one poller, one claim/lease, at-least-once, and consumers dedup (the core
 * `IdempotencyGuard`, or a UNIQUE latch like thread's consumed-message ledger). That is what makes a
 * TS→TS integration event survive a crash between the publish and the consumer. Until B3 this method
 * was an alias of `dispatch()`: nothing was written, so "the outbox will retry it" was simply false for
 * that direction.
 *
 * ORDERING, said out loud because the alias used to provide it by accident: an awaited in-memory
 * fan-out delivered A before B by construction. On the lane, order is `created_at` WITHIN one claim
 * batch and delivery is sequential — but a failed row does NOT hold back its successors here (this lane
 * does not group by owner; see `finalizeFailure`). A consumer that cannot tolerate "the later fact
 * arrived first" must say so at its own site.
 */
@injectable()
export class SqlExternalMediator extends ExternalMediator implements PollingService {
	private handlerMap = new Map<string, Handler[]>()
	private callbacks = new Set<EventCallback>()
	private namedCallbacks = new Map<string, Set<EventCallback>>()

	private timer: ReturnType<typeof setTimeout> | null = null
	private pollIntervalMs = POLL_MIN_MS
	private draining = false
	private stopped = true

	/**
	 * "Meu timer de poll está armado" — o menor sinal VERDADEIRO que já existia aqui.
	 *
	 * `stopped` e não `timer`: `stopped` nasce `true` e só é virado por `start()`/`stop()`, então
	 * `!stopped` é literalmente "start rodou e stop não" — correto ANTES do primeiro `start()`, que
	 * é a janela em que o probe de readiness pergunta. `timer` aqui só existe entre um
	 * `scheduleNext()` e o `drainOnce` seguinte.
	 */
	get running(): boolean {
		return !this.stopped
	}

	constructor(
		private driver: LibSqlDatabaseDriver,
		private domainEvents: DomainEventRepository,
	) {
		super()
	}

	override async start(): Promise<void> {
		this.stopped = false
		this.pollIntervalMs = POLL_MIN_MS
		this.scheduleNext()
	}

	override async stop(): Promise<void> {
		this.stopped = true
		if (this.timer) {
			clearTimeout(this.timer)
			this.timer = null
		}
		while (this.draining) await sleep(25)
	}

	async register(handler: Handler): Promise<void> {
		for (const name of handlerEventNames(handler)) {
			const existing = this.handlerMap.get(name)
			if (existing) existing.push(handler)
			else this.handlerMap.set(name, [handler])
		}
	}

	/** In-process fan-out ONLY — see the class docblock on why nothing is written here. */
	async dispatch(event: BaseEvent): Promise<void> {
		const handlers = this.handlerMap.get(event.name) ?? []
		const errors: Error[] = []
		for (const handler of handlers) {
			const outcome = await tryCatchAsync(async () => handler.execute(event))
			if (!outcome.success) errors.push(outcome.error)
		}
		this.notifyCallbacks(event.name, event)
		if (errors.length === 1) throw errors[0]
		if (errors.length > 1) {
			throw new AggregateError(errors, `${errors.length} of ${handlers.length} handlers failed for '${event.name}'`)
		}
	}

	/**
	 * PERSIST. Nothing else.
	 *
	 * The row is written inside the driver's write transaction (the only legitimate write path) and the
	 * method returns; `drainOnce` delivers it. The long note this docblock used to carry — about why the
	 * in-memory fan-out had to be awaited so `issue.opened` reached its consumer before
	 * `issue.completed` — described a property of the ALIAS, which is gone. What replaces it is the
	 * ORDERING paragraph on the class: intra-batch order, and consumers that dedup.
	 */
	async publish(event: BaseEvent): Promise<void> {
		// `publish` is typed on the widest event (the Mediator contract), but only integration events may
		// ride this lane: the row is scoped by the envelope `ownerId`, which a domain event does not have,
		// and an unscoped row is a row nothing can deliver. Fail loud rather than write it.
		if (!(event instanceof BaseIntegrationEvent)) {
			throw new BaseError<BaseInfrastructureErrors>(
				'INVALID_OUTBOX_PAYLOAD',
				`SqlExternalMediator.publish accepts integration events only — got '${event.name}'. Domain events are persisted by DomainEventRepository.save and dispatched on the api lane.`,
			)
		}
		// ONE documented boundary cast: `AnyIntegrationEvent` is the widest integration type the
		// persistence API accepts, but `instanceof` narrows to the class's DEFAULT schema instantiation,
		// not the concrete wire schema the row actually carries. A single widening cast is enough — the
		// two instantiations overlap structurally — and the `instanceof` IS the runtime proof.
		const integrationEvent = event as AnyIntegrationEvent
		await this.driver.transaction(tx => this.domainEvents.saveIntegrationEvent(integrationEvent, tx))
	}

	async execute<T extends Handler>(_handler: T['name'], _input: T['input']): Promise<T['output']> {
		throw new BaseError<BaseInfrastructureErrors>('NOT_IMPLEMENTED')
	}

	removeAllListeners(): void {
		this.handlerMap.clear()
		this.callbacks.clear()
		this.namedCallbacks.clear()
	}

	registerCallback(callback: EventCallback, eventName?: string): Unsubscribe {
		if (!eventName) {
			this.callbacks.add(callback)
			return () => this.callbacks.delete(callback)
		}
		const named = this.namedCallbacks.get(eventName) ?? new Set<EventCallback>()
		named.add(callback)
		this.namedCallbacks.set(eventName, named)
		return () => named.delete(callback)
	}

	/**
	 * One ingress pass: claim a leased batch of the `integration` lane, dispatch each row OUTSIDE
	 * the transaction, finalize. Returns how many rows were handled, so the poll loop can back off.
	 * Public because tests (and ops) drive it deterministically.
	 */
	async drainOnce(): Promise<number> {
		const names = [...this.handlerMap.keys()]
		// EMENDA O1 (B3) — THE CLAIM SERVES THE CALLBACKS TOO. A global callback (the SSE broadcaster is
		// one, registered once per process) consumes EVERY integration fact, not a named subset, so while
		// one exists the claim covers the WHOLE lane: handlers run for the names that have them,
		// `notifyCallbacks` fires for every claimed row. Without this, decision 5 ("the SSE fires FROM the
		// poller") would silently drop every TS-published fact with no backend consumer —
		// `thread.attached`, `issue.archived`, `issue.stop_resolved` — because they have no handler name
		// to be claimed by. Desired side effect: dormant rows are tombstoned after the broadcast instead
		// of accumulating unprocessed forever.
		const broadcasting = this.callbacks.size > 0
		// No handler AND no global callback ⇒ claim NOTHING. Claiming rows we cannot deliver would
		// burn their attempts budget and dead-letter the gateway's traffic. Mirrors the Go twin, and it
		// is what keeps a headless script (no SSE, no handlers) from tombstoning traffic it never reads.
		if (!broadcasting && names.length === 0) return 0

		const claimed = await this.claimBatch(broadcasting ? null : names)
		if (claimed.length === 0) return 0

		for (const row of claimed) {
			const outcome = await tryCatchAsync(() => this.deliver(row))
			if (outcome.success) await this.finalizeSuccess(row)
			else await this.finalizeFailure(row, String(outcome.error))
		}
		return claimed.length
	}

	private async deliver(row: ClaimedOutboxRow): Promise<void> {
		// Parse the payload OURSELVES, as raw TEXT, with the date reviver — see `reviveIsoDates`.
		const parsed = JSON.parse(row.payload ?? '{}', reviveIsoDates)
		// A flat Go wire struct is folded back into the nested BaseIntegrationEvent envelope; an
		// already-nested TS envelope passes through. Free defence, kept deliberately.
		const envelope = adaptWireEnvelope(parsed) as BaseEvent

		for (const handler of this.handlerMap.get(row.name) ?? []) {
			await handler.execute(envelope)
		}
		this.notifyCallbacks(row.name, envelope)
	}

	/**
	 * The claim, identical in protocol to the domain dispatcher's except for the lane and the
	 * handler-name filter. One `BEGIN IMMEDIATE` transaction, COMMITTED before anything is
	 * dispatched.
	 *
	 * `names === null` means "the whole lane" — the EMENDA O1 mode `drainOnce` selects while a global
	 * callback is registered. Any other value narrows to those names, as before.
	 *
	 * `attempts` is charged HERE, not on failure. An ingress event that kills the process has
	 * exactly the same failure mode as a domain one — and this is the lane carrying payloads the TS
	 * daemon never validated, so if either claimant were to go without a crash-loop ceiling it would
	 * be the wrong one. The poison sweep collects rows that burned the budget without ever
	 * finalizing; without it they are neither claimable nor terminal, i.e. invisible.
	 */
	private async claimBatch(names: string[] | null): Promise<ClaimedOutboxRow[]> {
		const now = Date.now()
		const token = crypto.randomUUID()
		const nameFilter = names
			? sql`AND ${outbox.name} IN (${sql.join(
					names.map(n => sql`${n}`),
					sql`, `,
				)})`
			: sql``

		return this.driver.transaction(async tx => {
			await tx.run(sql`
				UPDATE ${outbox}
				SET processed_at = ${now}, claimed_by = NULL, last_error = 'poison: exceeded attempts without finalize'
				WHERE ${outbox.source} = ${LANE}
					AND ${outbox.processedAt} IS NULL
					AND ${outbox.attempts} >= ${MAX_ATTEMPTS}
					AND ${outbox.leaseUntil} IS NOT NULL
					AND ${outbox.leaseUntil} < ${now}
			`)

			const due = await tx.all<{ id: string }>(sql`
				SELECT id
				FROM ${outbox}
				WHERE ${outbox.source} = ${LANE}
					AND ${outbox.processedAt} IS NULL
					AND ${outbox.attempts} < ${MAX_ATTEMPTS}
					AND (${outbox.leaseUntil} IS NULL OR ${outbox.leaseUntil} < ${now})
					${nameFilter}
				ORDER BY ${outbox.createdAt}
				LIMIT ${BATCH_SIZE}
			`)
			const ids = due.map(row => String(row.id))
			if (ids.length === 0) return []

			await tx
				.update(outbox)
				.set({
					claimedBy: token,
					leaseUntil: new Date(now + LEASE_MS),
					// Unqualified on purpose: single-table UPDATE, and `attempts + 1` is the whole point.
					attempts: sql`attempts + 1`,
				})
				.where(inArray(outbox.id, ids))

			// Raw re-read: `payload` must come back as TEXT so `deliver` can parse it with the
			// reviver. Reading it through the ORM's json column mode would parse it WITHOUT one.
			return tx.all<ClaimedOutboxRow>(sql`
				SELECT id, name, owner_id, payload, attempts
				FROM ${outbox}
				WHERE ${outbox.claimedBy} = ${token}
				ORDER BY ${outbox.createdAt}
			`)
		})
	}

	/** Tombstone. The row is kept — a deleted id is a re-insertable id on the Go side. */
	private async finalizeSuccess(row: ClaimedOutboxRow): Promise<void> {
		await this.driver.transaction(tx => tx.update(outbox).set({ processedAt: new Date(), claimedBy: null }).where(eq(outbox.id, row.id)))
	}

	/**
	 * Below the ceiling: record the error and NOTHING else — the lease is retained on purpose (that
	 * IS the 30s backoff) and `attempts` was charged at claim time. At or above it: dead-letter.
	 * There is no skip branch here: this lane does not group by owner (nor does the Go twin), so
	 * owner-sequential ordering and its skip belong to the `api` dispatcher alone.
	 */
	private async finalizeFailure(row: ClaimedOutboxRow, error: string): Promise<void> {
		const deadLettered = Number(row.attempts) >= MAX_ATTEMPTS
		await this.driver.transaction(tx =>
			tx
				.update(outbox)
				.set(deadLettered ? { lastError: error, processedAt: new Date(), claimedBy: null } : { lastError: error })
				.where(eq(outbox.id, row.id)),
		)
	}

	private notifyCallbacks(eventName: string, event: BaseEvent): void {
		for (const cb of this.callbacks) cb(event)
		const named = this.namedCallbacks.get(eventName)
		if (named) for (const cb of named) cb(event)
	}

	private scheduleNext(): void {
		if (this.stopped) return
		this.timer = setTimeout(() => void this.poll(), Math.min(this.pollIntervalMs, POLL_MAX_MS))
		this.timer.unref?.()
	}

	private async poll(): Promise<void> {
		if (this.stopped || this.draining) return
		this.draining = true
		try {
			const handled = await this.drainOnce()
			this.pollIntervalMs = handled > 0 ? POLL_MIN_MS : Math.min(this.pollIntervalMs * POLL_BACKOFF_FACTOR, POLL_MAX_MS)
		} catch (error) {
			console.error('SqlExternalMediator ingress poll failed:', error)
			this.pollIntervalMs = Math.min(this.pollIntervalMs * POLL_BACKOFF_FACTOR, POLL_MAX_MS)
		} finally {
			this.draining = false
			this.scheduleNext()
		}
	}
}

function sleep(ms: number): Promise<void> {
	return new Promise(resolve => setTimeout(resolve, ms))
}
